src/mongo_logger.ts

```diff
@@ -0,0 +1,214 @@
+  MONGODB_LOG_CONNECTION?: string;
+  /** Default severity level to be if any of the above are unset */
+  MONGODB_LOG_ALL?: string;
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  MONGODB_LOG_MAX_DOCUMENT_LENGTH?: string;
+  /** Destination for log messages. Must be 'stderr', 'stdout'. Defaults to 'stderr'. */
+  MONGODB_LOG_PATH?: string;
+}
+
+/** @internal */
+export interface MongoLoggerMongoClientOptions {
+  /** Destination for log messages */
+  mongodbLogPath?: 'stdout' | 'stderr' | Writable;
+}
+
+/** @internal */
+export interface MongoLoggerOptions {
+  componentSeverities: {
+    /** Severity level for command component */
+    command: SeverityLevel;
+    /** Severity level for topology component */
+    topology: SeverityLevel;
+    /** Severity level for server selection component */
+    serverSelection: SeverityLevel;
+    /** Severity level for connection component */
+    connection: SeverityLevel;
+    /** Default severity level to be used if any of the above are unset */
+    default: SeverityLevel;
+  };
+
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  maxDocumentLength: number;
+  /** Destination for log messages. */
+  logDestination: Writable;
+}
+
+/**
+ * Parses a string as one of SeverityLevel
+ *
+ * @param s - the value to be parsed
+ * @returns one of SeverityLevel if value can be parsed as such, otherwise null
+ */
+function parseSeverityFromString(s?: string): SeverityLevel | null {
+  const validSeverities: string[] = Object.values(SeverityLevel);
+  const lowerSeverity = s?.toLowerCase();
+
+  if (lowerSeverity != null && validSeverities.includes(lowerSeverity)) {
+    return lowerSeverity as SeverityLevel;
+  }
+
+  return null;
+}
+
+/**
+ * Parses a string to be a number greater than or equal to 0 for maxDocumentLength.
+ *
+ * @param s - the value to be parsed
+ * @returns the int value parsed or 1000 if the value could not be parsed
+ */
+function parseMaxDocumentLength(s?: string): number {
+  try {
+    const maxDocumentLength = getUint('MONGODB_LOG_MAX_DOCUMENT_LENGTH', s);
+    return maxDocumentLength;
+  } catch {
+    return 1000;
+  }
+}
+
+/**
+ * resolves the MONGODB_LOG_PATH and mongodbLogPath options from the environment and the
+ * mongo client options respectively.
+ *
+ * @returns the Writable stream to write logs to
+ */
+function resolveLogPath(
+  { MONGODB_LOG_PATH }: MongoLoggerEnvOptions,
+  {
+    mongodbLogPath
+  }: {
+    mongodbLogPath?: unknown;
+  }
+): Writable {
+  const isValidLogDestinationString = (destination: string) =>
+    ['stdout', 'stderr'].includes(destination.toLowerCase());
+  if (typeof mongodbLogPath === 'string' && isValidLogDestinationString(mongodbLogPath)) {
+    return mongodbLogPath.toLowerCase() === 'stderr' ? process.stderr : process.stdout;
+  }
+
+  if (typeof mongodbLogPath === 'object' && mongodbLogPath instanceof Writable) {
```

Comment:
It's obviously not wrong for us to check `instanceof Writable` here, but we may want to consider trusting the user has passed us something with a compatible interface that isn't an instanceof a Node.js stream. This may be easier to remove once we have a clear idea about the exact APIs we need to exist (ex. `.write()`, `.flush()` etc. w.e. is minimally necessary)

Reply:
I had this thought, too, but my original comment is now lost somewhere here or in the depths of slack; thanks for catching that - we should leave a `TODO(NODE-xxxx): check for minimal interface instead of instanceof Writable` with the ticket that will actually implement the logging methods

Reply:
@baileympearson Instead of the empty [NODE-4886](https://jira.mongodb.org/browse/NODE-4886), can we instead update the description and AC of [NODE-4813](https://jira.mongodb.org/browse/NODE-4813)?